src/dep_graph.rs

```diff
@@ -0,0 +1,110 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use cddl::ast::*;
+
+pub fn topological_rule_order<'a>(rules: &'a Vec<&'a Rule<'a>>) -> Vec<&'a Rule<'a>> {
+    let mut adj_list = BTreeMap::new();
+    for cddl_rule in rules.iter() {
+        let (ident, refs) = find_references(cddl_rule);
+        adj_list.insert(ident.ident, (*cddl_rule, refs));
+    }
+    let mut unvisited = adj_list.iter().map(|(k, _v)| *k).collect::<BTreeSet<&str>>();
+    let mut topo_order = Vec::new();
+    let mut processing: BTreeSet<&'a str> = BTreeSet::new();
+    while let Some(u) = unvisited.iter().next().map(|u| *u) {
+        dfs_visit(&mut topo_order, &mut unvisited, &mut processing, &adj_list, u);
+    }
+    topo_order
+}
+
+fn dfs_visit<'a>(
+    topo_order: &mut Vec<&'a Rule<'a>>,
+    unvisited: &mut BTreeSet<&'a str>,
+    processing: &mut BTreeSet<&'a str>,
+    adj_list: &BTreeMap<&'a str, (&'a cddl::ast::Rule<'a>, Vec<&'a Identifier<'a>>)>,
+    u: &'a str) {
+    processing.insert(u);
+    let (rule, neighbors) = adj_list.get(u).unwrap();
+    for v in neighbors.iter() {
+        if processing.contains(v.ident) {
+            eprintln!("Recursive type: '{}' / '{}' - code will possibly need to be edited by hand to use Box/etc", u, v);
```

Comment:
At some point we could potentially figure out if this could be automated, but I don't think it's super high priority right now and the hand edit to fix it is simple (just replace `T` with `Box<T>`). When the recursive type is in another heap-allocated structure like `Vec`/`BTreeMap`/etc no changes are needed, it's just when it's stack-allocated directly that you run into issues.

At least in the split-up `babbage.cddl` this ignoring cycles but continuing on with the topological sort seemed to be fine.

Reply:
Maybe let's crate a Github issue to track this? We can leave it open for now

Reply:
I created one here: #105